hpx/parallel/executors/parallel_executor_aggregated.hpp

```diff
@@ -0,0 +1,356 @@
+#include <functional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace parallel { namespace execution
+{
+    ///////////////////////////////////////////////////////////////////////////
+    /// A \a parallel_executor_aggregated creates groups of parallel execution
+    /// agents that execute in threads implicitly created by the executor. This
+    /// executor prefers continuing with the creating thread first before
+    /// executing newly created threads.
+    ///
+    /// This executor conforms to the concepts of a TwoWayExecutor,
+    /// and a BulkTwoWayExecutor
+
+    template <typename Policy = hpx::launch::async_policy>
+    struct parallel_policy_executor_aggregated
+    {
+        /// Associate the parallel_execution_tag executor tag type as a default
+        /// with this executor.
+        typedef parallel_execution_tag execution_category;
+
+        /// Associate the static_chunk_size executor parameters type as a default
+        /// with this executor.
+        typedef static_chunk_size executor_parameters_type;
+
+        /// Create a new parallel executor
+        HPX_CONSTEXPR explicit parallel_policy_executor_aggregated() {}
+
+        /// \cond NOINTERNAL
+        bool operator==(parallel_policy_executor_aggregated const& rhs) const
+            noexcept
+        {
+            return true;
+        }
+
+        bool operator!=(parallel_policy_executor_aggregated const& rhs) const
+            noexcept
+        {
+            return false;
+        }
+
+        parallel_policy_executor_aggregated const& context() const noexcept
+        {
+            return *this;
+        }
+        /// \endcond
+
+        /// \cond NOINTERNAL
+
+        // TwoWayExecutor interface
+        template <typename F, typename ... Ts>
+        hpx::future<void>
+        async_execute(F && f, Ts &&... ts) const
+        {
+            return hpx::detail::async_launch_policy_dispatch<Policy>::call(
+                Policy{}, std::forward<F>(f), std::forward<Ts>(ts)...);
+        }
+
+        // NonBlockingOneWayExecutor (adapted) interface
+        template <typename F, typename ... Ts>
+        static void post(F && f, Ts &&... ts)
+        {
+            hpx::util::thread_description desc(f,
+                "hpx::parallel::execution::parallel_executor_aggregated::post");
+
+            detail::post_policy_dispatch<Policy>::call(
+                desc, Policy{}, std::forward<F>(f), std::forward<Ts>(ts)...);
+        }
+
+    private:
+        /// \cond NOINTERNAL
+        template <typename F, typename ... Ts>
+        struct wrapped_func
+        {
+            template <typename Value>
+            void operator()(
+                lcos::local::counting_semaphore& sem, Value const& value)
+            {
+                // invoke user function
+                hpx::util::invoke_fused(f_,
+                    hpx::util::tuple_cat(
+                        hpx::util::forward_as_tuple(value), ts_));
```

Comment:
What happens when f_ throws an exception? The semaphore will never be signaled.

Reply:
Excellent catch.